### PR TITLE
[IMP] l10n_in_edi: add view e-invoice in portal action

### PR DESCRIPTION
**Pull request by habh-odoo**
Before this commit
====================
There is no action to view the e-Invoice in Portal.

After this commit
====================
Added an action to view the e-Invoice in Portal.

task-3336590

### CODE REVIEW

File: addons/l10n_in_edi/controllers/main.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+import json
+from odoo.addons.l10n_in_edi import utils as l10n_in_edi_utils
+import werkzeug.exceptions
+
+class ViewEinvoiceController(http.Controller):
+
+    @http.route(['/l10n_in_edi/vieweinvoice/<int:invoice_id>'], type='http')
+    def controller_view_einvoice(self, invoice_id):
+        move = http.request.env['account.move'].search([('id', '=', invoice_id)])
+        json_data = move and move._get_l10n_in_edi_response_json() or False
```
Comment (jbw-odoo) [inline addons/l10n_in_edi/controllers/main.py:14]:
```suggestion
        json_data = move and move._get_l10n_in_edi_response_json()
```

File: addons/l10n_in_edi/utils.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+def verify_signed_invoice(json_data):
+    url = "https://einvoice1.gst.gov.in/Others/VSignedInvoice"
+    get_res = requests.get(url)
+    cookies = get_res.cookies.get_dict()
+    html = get_res.text
+    html = lxml.html.fromstring(html)
+    requestverificationtoken = html.xpath('//input[@name="__RequestVerificationToken"]/@value')[0]
+    payload = {
+        'submit': 'Upload',
+        '__RequestVerificationToken': requestverificationtoken
```
Comment (jbw-odoo) [inline addons/l10n_in_edi/utils.py:16]:
```suggestion
        '__RequestVerificationToken': requestverificationtoken,
```

File: addons/l10n_in_edi/utils.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    }
+    files = [
+        ('file', ('json_data.json', json_data.encode(), 'application/json'))
+    ]
```
Comment (jbw-odoo) [inline addons/l10n_in_edi/utils.py:20]:
```suggestion
    files = [('file', ('json_data.json', json_data.encode(), 'application/json'))]
```

File: addons/l10n_in_edi/controllers/main.py (hunk 1 above)
Comment (habh-odoo) [reply addons/l10n_in_edi/controllers/main.py]: done

File: addons/l10n_in_edi/utils.py (hunk 2 above)
Comment (habh-odoo) [reply addons/l10n_in_edi/utils.py]: done

File: addons/l10n_in_edi/utils.py (hunk 3 above)
Comment (habh-odoo) [reply addons/l10n_in_edi/utils.py]: done

**Comment by jbw-odoo** [comment]
Can you also add the task or ticket id if there is one in your commit message  ?

**Comment by habh-odoo** [comment]
@jbw-odoo added task id.

File: addons/l10n_in_edi/utils.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    html_text = get_res.text
+    html = lxml.html.fromstring(html_text)
+    try:
+        requestverificationtoken = html.xpath('//input[@name="__RequestVerificationToken"]/@value')[0]
+    except:
+        return html_text
+    payload = {
+        'submit': 'Upload',
+        '__RequestVerificationToken': requestverificationtoken,
+    }
```
Comment (jbw-odoo) [inline addons/l10n_in_edi/utils.py:20]:
```suggestion
    requestverificationtoken = html.xpath('//input[@name="__RequestVerificationToken"]/@value')
    if not requestverificationtoken:
        return html_text
    payload = {
        'submit': 'Upload',
        '__RequestVerificationToken': requestverificationtoken[0],
    }
```
If the try except is there only for the case where ``html.xpath(...)`` would return an empty list.

Comment (habh-odoo) [reply addons/l10n_in_edi/utils.py]: done

**Comment by jbw-odoo** [comment]
I wanted to test it with an demo data invoice but got "Forbidden

You don't have the permission to access the requested resource. It is either read-protected or not readable by the server."

Same when i tested on runbot : 
https://39574220-15-0-all.runbot151.odoo.com/l10n_in_edi/vieweinvoice/4

**Comment by jbw-odoo** [comment]
> Hi @jbw-odoo, this occurs when there is no Indian E-invoice. I have added a check to raise User Error in this case.

Hi @habh-odoo . Then you should maybe hide the action if there is no e-invoice ? Better than display an error afterwards. Ask the PO ? And on that note I realized your task is not in the r&d review stage yet.
I suggest you let the PO check your task first, and then, when it's ready to be code reviewed you can ping me 😀

**Comment by jbw-odoo** [comment]
>  Then you should maybe hide the action if there is no e-invoice ? 

If it's possible with an action 🤔 ?

**Comment by habh-odoo** [comment]
hi @jbw-odoo, it is not possible to hide the action.

**Comment by jva-odoo** [comment]
move to https://github.com/odoo/odoo/pull/127716

**Pull request closed without merging**